### PR TITLE
Add gift promotion

### PR DESCRIPTION
**Pull request by IKarbowiak**
Introduce a new `gift` reward in `Promotion Rule`.

- The `PromotionRule` of `gift` reward type has list of available `gifts` that are product variants instances, which can be added to checkout as a gift
- When the gift promotion rule applies, the variant from available variants with the top price is chosen as a gift, and a new gift line is created on checkout, marked with `isFlag` set to `True`.
- There is a limit of 500 maximum assigned gifts to `PromotionRule`

⏲️ 
Checkout calculations were measured for 100 order promotion rules (the maximum number of rules currently) and with 500 assigned gifts to each rule and the calculation time is below 1s.

Resolves: https://github.com/saleor/saleor/issues/12913

# Impact

- [x] New migrations
- [x] New/Updated API fields or mutations
- [ ] Deprecated API fields or mutations
- [ ] Removed API types, fields, or mutations

# Docs

- [ ] Link to documentation: https://github.com/saleor/saleor-docs/pull/1080

# Pull Request Checklist

- [x] Privileged queries and mutations are either absent or guarded by proper permission checks
- [x] Database queries are optimized and the number of queries is constant
- [x] Database migrations are either absent or optimized for zero downtime
- [x] The changes are covered by test cases
- [x] All new fields/inputs/mutations have proper labels added (`ADDED_IN_X`, `PREVIEW_FEATURE`, etc.)
- [x] All migrations have proper dependencies
- [x] All indexes are added concurrently in migrations
- [x] All RunSql and RunPython migrations have revert option defined

### CODE REVIEW

File: saleor/checkout/complete_checkout.py [hunk 1]
```diff
@@ -355,6 +352,15 @@ def _create_order_line_discounts(
     return line_discounts
 
 
+def _get_sale_id(line_discounts: list[OrderLineDiscount]):
+    sale_id = None
+    for discount in line_discounts:
+        if discount.type == DiscountType.PROMOTION:
+            if rule := discount.promotion_rule:
+                sale_id = get_sale_id(rule.promotion)
+    return sale_id
```
Comment (tomaszszymanski129) [inline saleor/checkout/complete_checkout.py:361]: from above usage, I understand this is meant to return one possible sale, maybe we can simplify it?
```suggestion
def _get_sale_id(line_discounts: list[OrderLineDiscount]):
    for discount in line_discounts:
        if discount.type == DiscountType.PROMOTION:
            if rule := discount.promotion_rule:
                return get_sale_id(rule.promotion)
```

File: saleor/graphql/checkout/mutations/checkout_lines_update.py [hunk 2]
```diff
@@ -221,3 +224,28 @@ def _get_variants_from_lines_input(cls, lines: list[dict]) -> list[ProductVarian
...
+            str(line_info.line.id) for line_info in lines_info if line_info.line.is_gift
+        ]
+        gift_lines_to_update = [
+            line for line in lines_input if line.line_id in existing_gift_ids
+        ]
+        if gift_lines_to_update:
+            global_ids = [
+                graphene.Node.to_global_id("CheckoutLine", line.line_id)
+                for line in gift_lines_to_update
+            ]
```
Comment (tomaszszymanski129) [inline saleor/graphql/checkout/mutations/checkout_lines_update.py:242]:
```suggestion
        if gift_lines_to_update := [
            line for line in lines_input if line.line_id in existing_gift_ids
        ]:
            global_ids = [
                graphene.Node.to_global_id("CheckoutLine", line.line_id)
                for line in gift_lines_to_update
            ]
```

File: saleor/graphql/discount/mutations/promotion/validators.py [hunk 3]
```diff
@@ -57,6 +60,47 @@ def clean_promotion_rule(
...
+        gift_ids = {gift.id for gift in cleaned_input["gifts"]}
+    else:
+        # this part is only for PromotionRuleUpdate mutation
+        # so the gifts will be fetched once
+        gift_ids = (
+            {gift.id for gift in instance.gifts.all()}
+            | {gift.id for gift in cleaned_input.get("add_gifts", [])}
+        ) - {gift.id for gift in cleaned_input.get("remove_gifts", [])}
+
+    return gift_ids
```
Comment (tomaszszymanski129) [inline saleor/graphql/discount/mutations/promotion/validators.py:80]: This is complicated already, maybe we can get rid of the variable? What both cases do is return list of ids at 1st glance so we don't loose much by not assigning those to named variable.
```suggestion
    if "gifts" in cleaned_input:
        return {gift.id for gift in cleaned_input["gifts"]}
    else:
        # this part is only for PromotionRuleUpdate mutation
        # so the gifts will be fetched once
        return (
            {gift.id for gift in instance.gifts.all()}
            | {gift.id for gift in cleaned_input.get("add_gifts", [])}
        ) - {gift.id for gift in cleaned_input.get("remove_gifts", [])}
```

**Review by tomaszszymanski129 — APPROVED** [review]

File: saleor/graphql/discount/mutations/promotion/validators.py (hunk 3 above)
Comment (IKarbowiak) [reply saleor/graphql/discount/mutations/promotion/validators.py]: I removed `gift_ids` variable and created a variable for each of sets to make the calculation more clear. I hope it helps

**Review by 8r2y5 — APPROVED** [review]

**Review by korycins — APPROVED** [review]

**Review by fowczarek — APPROVED** [review]

**Review by kadewu — APPROVED** [review]

**Pull request merged**